### ISSUE TITLE
Simplify statistics SPI

### ISSUE THREAD

**Issue by arhimondr** [prestodb/presto#11436]
- Remove `RangeColumnStatistics`

`RangeColumnStatistics` assumes fixed ranges of values. Most likely we would like to implement histograms by having non fixed ranges of values that have similar carnality. And those ranges may be different for distinct values, values count overall, and data size. Also fraction of nulls doesn't make much sense in terms of histograms. Since null is a single value. And we know the exact count of it.

- Represent `lowValue` and `highValue` as `double` in the SPI. 

Although I don't feel strong about this, it seems like a good simplification for a short term. It should simplify unit testing for the MetastoreHiveStatisticsProvider a lot. Also having MIN/MAX as double elliminated the need of schema migration (there's no need to cap the values if the column type changed from BIGINT to INT).

Our statistic model cannot leverage other min/max values than simple numeric. And those are represented as `double` in the stats model. It is unclear at what step are we going to introduce MIN/MAX for strings. It is unclear whether we need this at all. And it is also unclear if we would like to introduce it before we have Histograms. In case we have histograms - the MIN/MAX would be extra. Because this information can be obtained from the Histograms itself.

**Comment by arhimondr** [comment]
CC: @mbasmanova @rschlussel @kokosing @findepi @sopel39

**Comment by findepi** [comment]
> Also having MIN/MAX as double elliminated the need of schema migration (there's no need to cap the values if the column type changed from BIGINT to INT).

Please explain. Can i have an INT column with max over 2^31, eg 2^57?

**Comment by kokosing** [comment]
> RangeColumnStatistics ... assumes fixed ranges of values. Also fraction of nulls doesn't make much sense in terms of histograms.

IIRC `com.facebook.presto.spi.statistics.RangeColumnStatistics#fraction` is a fraction of of data (not nulls), so currently it is `1 - nullsFraction` in case of single range. Having that said the quote above is incorrect. Or I misunderstood what you mean by `fixed ranges of values`.

> Represent lowValue and highValue as double in the SPI.

That would force connectors to make translation from object to double. I would prefer to avoid that as real value is easy to understand and to implement in 3rd party connectors, double representation becomes non trivial.

> It should simplify unit testing for the MetastoreHiveStatisticsProvider a lot.

I think there are plenty of ways to handle that other than changing the SPI. IMO SPI should be generic, it should not reflect 1:1 engine capabilities, but rather future potential of engine capabilities. 

Notice that SPI may return multiple `RangeColumnStatistics`s for single column, so it is already prepared for histograms, but engine does not support that fully yet.

**Comment by arhimondr** [comment]
TLDR:

- Everything around Histogramms is very unclear
- Optimizer doesn't understand any other MIN/MAX but double. 
- It is easier for connectors to covert internal MIN/MAX representation to Double

> IIRC com.facebook.presto.spi.statistics.RangeColumnStatistics#fraction is a fraction of of data (not nulls), so currently it is 1 - nullsFraction in case of single range. Having that said the quote above is incorrect.

Yes, you are right. I didn't get this at the first place. Since we don't have more than a single range, the part with `1 - nullsFraction` was confusing.

> Or I misunderstood what you mean by fixed ranges of values

Currently in the `RangeColumnStatistics`, besides `fraction`, we have two additional metrics: `dataSize`, `distinctValuesCount`. By fixed range i meant that for a specific range we have multiple metrics. The only way i can see we can compute them, is to have the range fixed, and than group the values per pre-defined ranges.

Setting the ranges can be done:

- Statically. For example by splitting the INTEGER range into number of buckets. This is suboptimal, as the data might not be evenly distributed. And it is even worse when speaking in terms of DOUBLE or STRING.
- Dynamically. Set the number of desired buckets, scan the data, get the ranges with a similar cardinality.

Choosing the second approach is preferable. But when choosing the second approach it is not guaranteed that the NDV distribution will be the same as all values distribution. And it is unclear how to normalize it and merge.

What is even more important, we are not even sure we would ever want to collect histograms over the NDV. As it is unclear how can it be used. CC: @rschlussel 

Also there is a question of efficiency for collecting histograms. Ideally we would like all the statistics to be collected in a streeming way, without additional memory overhead. There is an efficient algoritms for building histograms in a streaming way by detecting fixed number of ranges that include similar number of values. Here is the paper: https://www.cis.upenn.edu/~sudipto/mypapers/histogram_first.pdf But as i understand works only for numeric data, and doesn't work for STRINGS. And i'm not sure if there is an efficient streaming algorithm for collecting histogramms for strings. That's why it is very unclear if we would ever want to have histogramms over any other columns than numeric. As it may be just to expensive to compute such statistics. 

Based on this thoughts and assumptions, by my opinion, the most optimal and simple histogram representation would be the `List<Double>`, where first element is min value, last element is the max value, and values in the middle represent equal cardinality buckets. So the `ColumnStatistics` may have 

```
List<Double> overallCardinallityHistogramm
List<Double> ndvCardinallityHistogram -- if we ever wanted to have something like that
```

But speaking about histogramms is just tooo futuristic. Our statistics model can't understand them, and honestly this is still unclear in what use cases would histograms help us with, and whether the benefits from using histogramms will cover the computation costs. That's why i strongly believe that having the SPI more complicated because of some very very futuristic vision that is very blurred is worth it.

> That would force connectors to make translation from object to double. I would prefer to avoid that as real value is easy to understand and to implement in 3rd party connectors, double representation becomes non trivial.

Currently all connectors that provide statistics (Hive/Tpch/Tpcds) have the min and max values strongly typed, and it is way easier to just convert them to `double`, than to have to deal with untyped `Object`, always wondering "What's the value is in there"??

https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java#L268
https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java#L166
https://github.com/prestodb/presto/blob/master/presto-tpcds/src/main/java/com/facebook/presto/tpcds/statistics/TpcdsTableStatisticsFactory.java#L89
https://github.com/prestodb/presto/blob/master/presto-tpch/src/main/java/com/facebook/presto/tpch/TpchMetadata.java#L392

> I think there are plenty of ways to handle that other than changing the SPI

Yes. But the simpler - the better.

> IMO SPI should be generic, it should not reflect 1:1 engine capabilities

Aggreed. But from the other side we should keep the SPI simple. The optimizer doesn't use any other values but double? Why do we need to complicate the SPI if we don't need the complication anyway?

**Comment by rschlussel** [comment]
I understand the value of having a fairly stable spi, but agree with @arhimondr that it's premature to design for histograms given that there are lots of different ways to implement them and we haven't really thought through the design for it and what we'd really want from the connectors. If it didn't complicate things too much, I'd say leave it in there, and then we can change it if we really need, but since it does, I'd rather keep it simpler.  I don't have strong feelings about double vs. generic.

**Comment by findepi** [comment]
> Remove RangeColumnStatistics

i am ok

> Represent lowValue and highValue as double in the SPI.

i would choose to go this way, but i am ok either way.

**Comment by kokosing** [comment]
@arhimondr Sounds good, thanks for explanation.

**Comment by sopel39** [comment]
> What is even more important, we are not even sure we would ever want to collect histograms over the NDV. As it is unclear how can it be used. CC: @rschlussel

By histograms you mean exact mapping from `value -> number of occurrences` or NDVs value per statistics range?

Keep in mind that NDVs are essential to various stats computation (e.g: `join, expr1 = exprs2, range intersection`). They seem integral to range information like cardinality and min/max. It would be strange to perform histogram computations and additionally have some "out-of-bound" global NDV information which would be difficult to estimate when doing range computations. That being said, we currently use only NDV information without min/max stats when doing join estimates. This was because such heuristics performed better when no column correlation information is available.

> But as i understand works only for numeric data, and doesn't work for STRINGS. And i'm not sure if there is an efficient streaming algorithm for collecting histogramms for strings.

For strings it might work better to just store exact string values with top cardinalates (maybe algorithm like: https://www.quora.com/What-algorithm-is-used-for-top-k-frequent-items-in-data-streams, https://arxiv.org/pdf/1408.1847.pdf could help estimate those). This way we could better conditions like "people is in NY state". 

One not that instead of pushing for histograms in a first place, we could push for simple column correlation estimate (e.g: https://en.wikipedia.org/wiki/Pearson_correlation_coefficient). This would greatly improve stats estimation for dimensional tables, where table id's are linearly correlated with dimension columns.

**Comment by arhimondr** [comment]
> Keep in mind that NDVs are essential to various stats computation (e.g: join, expr1 = exprs2, range intersection). They seem integral to range information like cardinality and min/max. It would be strange to perform histogram computations and additionally have some "out-of-bound" global NDV information which would be difficult to estimate when doing range computations. That being said, we currently use only NDV information without min/max stats when doing join estimates. This was because such heuristics performed better when no column correlation information is available.

That makes sense. But again, if we would like to compute histograms over the NDV i would still go with "same cardinality buckets" representation, that doesn't fit well to the current SPI design.

> For strings it might work better to just store exact string values with top cardinalates (maybe algorithm like: https://www.quora.com/What-algorithm-is-used-for-top-k-frequent-items-in-data-streams, https://arxiv.org/pdf/1408.1847.pdf could help estimate those). This way we could better conditions like "person is in NY state".

I need to read this paper. But if we decided to compute something like this - we would have to extend the SPI anyway.

> One not that instead of pushing for histograms in a first place, we could push for simple column correlation estimate (e.g: https://en.wikipedia.org/wiki/Pearson_correlation_coefficient). This would greatly improve stats estimation for dimensional tables, where table id's are linearly correlated with dimension columns.

Computing correlation between all the columns is very expensive, as it is needed to compute N*N correlations, where N - is the number of column in the table. To be able to compute correlation, we will need a mechanism how can we mark the columns to compute correlations for. As computing correlation for all the combination of columns will be just too expensive. There are some thoughts of how can we instrument optimizer to produce such markers automatically. But i think this goes way beyond the discussion in this issue.

**Comment by sopel39** [comment]
> But again, if we would like to compute histograms over the NDV i would still go with "same cardinality buckets" representation

I think that Bucket's cardinality shouldn't affect NDVs computation.

> Computing correlation between all the columns is very expensive, as it is needed to compute N*N correlations, where N - is the number of column in the table. To be able to compute correlation, we will need a mechanism how can we mark the columns to compute correlations for. As computing correlation for all the combination of columns will be just too expensive.

I'm not sure that computing NxN linear correlations would be that expensive. Linear correlation requires computing covariange matrix and there is tons of work how to do so efficiently (in a streaming way or by estimation).
Also we might not need to compute all N*N correlations but correlations between some "leading" column (e.g: column id) and rest of columns. "Leading" column (id column) could be set via some table property. Keep not that we could estimate correlation between two columns via "leading" column.
I agree that computing correlations might be too expensive when doing `INSERT`, but for running `ANALYZE` that should be fine.
